Add mochad light component

mtreinish opened the pull request:
## Description:
This commit adds a new light component for mochad to control x10 dimmers/lights with
mochad.


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2976

## Example entry for `configuration.yaml` (if applicable):
```yaml
light:
  - platform: mochad
   devices:
     - address: a1
     - address: a5
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

mtreinish committed:
Add mochad light component

This commit adds a new component to control x10 dimmers/lights with
mochad.

fabaff commented on homeassistant/components/light/mochad.py:
```diff
@@ -0,0 +1,94 @@
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ADDRESS = 'address'
+CONF_DEVICES = 'devices'
```
This is already present in `const.py`.

mtreinish replied:
Done

fabaff commented on homeassistant/components/light/mochad.py:
```diff
@@ -0,0 +1,94 @@
+Contains functionality to use a X10 dimmer over Mochad.
+
+For more details about this platform, please refer to the documentation at
+https://home.assistant.io/components/light.mochad
```
Tailing `/` is missing.

mtreinish replied:
Done

fabaff commented on homeassistant/components/light/mochad.py:
```diff
@@ -0,0 +1,94 @@
+CONF_ADDRESS = 'address'
+CONF_DEVICES = 'devices'
+
+PLATFORM_SCHEMA = vol.Schema({
```
Please extend the schema of the platform.

```python
PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
...
```

mtreinish replied:
Done

fabaff commented on homeassistant/components/light/mochad.py:
```diff
@@ -0,0 +1,94 @@
+
+        self._controller = ctrl
+        self._address = dev[CONF_ADDRESS]
+        self._name = dev.get(CONF_NAME, 'x10_light_dev_%s' % self._address)
```
Please use the "new" string formatting (`().format()`).

mtreinish replied:
Done

fabaff commented on homeassistant/components/light/mochad.py:
```diff
@@ -0,0 +1,94 @@
+    CONF_DEVICES: [{
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Required(CONF_ADDRESS): cv.x10_address,
+        vol.Optional('comm_type'): cv.string,
```
Suggestion make `'comm_type'` a constant as it's used later too.

mtreinish replied:
Done, I also updated the mochad switch component which has the same constant. I did the same for address which will also be common between all mochad components.

fabaff commented on homeassistant/components/light/mochad.py:
```diff
@@ -0,0 +1,94 @@
+        self.device = device.Device(ctrl, self._address,
+                                    comm_type=self._comm_type)
+        self._brightness = 0
+        self._state = self._get_device_status()
```
Use `def update()` to get the device status.

mtreinish replied:
This is just being used to set the initial state, we actually don't want to rely on this method for state updates after the devices is initialized. Mochad doesn't actually know the devices state it's just tracking in memory the last command it received. Just like when I implemented the mochad switch in https://github.com/home-assistant/home-assistant/pull/3970 relying on mochad is for device state just makes things go out of sync and cause weird behavior. It was much more reliable to have home-assistant track the device state.

fabaff reviewed (changes requested):
Just some quick comments

mtreinish committed:
Create comm_type and address constants

The comm_type and address conf constants are shared between all mochad
devices because they are required information used for configuring a
device. This commit moves the definition into const.py so they're
consistent between all component types.

pvizeli commented on homeassistant/const.py:
```diff
@@ -69,6 +70,7 @@
 CONF_BRIGHTNESS = 'brightness'
 CONF_CODE = 'code'
 CONF_COLOR_TEMP = 'color_temp'
+CONF_COMM_TYPE = 'comm_type'
```
Put that into your platforms

mtreinish replied:
Done, I moved it into components/mochad.py since it will be a shared config field for all the mochad devices.

pvizeli reviewed (changes requested):
Looks good after a small change.

pvizeli commented on homeassistant/components/light/mochad.py:
```diff
@@ -0,0 +1,94 @@
+    def turn_on(self, **kwargs):
+        """Send the command to turn the light on."""
+        self._brightness = kwargs.get(ATTR_BRIGHTNESS, 255)
+        self.device.send_cmd('xdim %s' % self._brightness)
```
use `"my string {}".format(bla)`

mtreinish replied:
Done

pvizeli commented on homeassistant/components/light/mochad.py:
```diff
@@ -0,0 +1,94 @@
+    return True
+
+
+class MochadLight(Light):
```
How work your class? Do you need poll data, add a `update()` function and read the real state of that device instead only on `__init__` and remove `read_data` from control functions. Looks into `Entity` class to see the available interface for devcie types.

mtreinish replied:
Please see the earlier comments in this PR on this. We don't want to rely on the mochad daemon to track the device status because the x10 devices are 1 way and it doesn't actually know the state. (it just stores a record of the last command sent) It also causes a synchronization issue because the daemon is a serial interface. Having home assistant track the device state is much more reliable. The only reason I poll for the status on __init__ is to take a better guess at the initial state.

The read_data() call in all the control functions is just for synchronization, it is waiting for the ack from the mochad daemon on the socket interface.

pvizeli replied:
If we have no access to real state, you need overwrite this interface: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity.py#L147

You can also make a poll device and set `SCAN_INTERVAL=timedelta(hour=1)` so we have a minimal poll on the device it self and use `add_devices(..., True)`

mtreinish replied:
Done, I added the assumed_state property to the class.

pvizeli commented:
Good work

gwijsman commented:
hi guys, thanks for developing this great component. I have encounterd an issue with the multi threading on switches and lights. do you want me to add my issue log to his issue or issue a new one?

pvizeli commented:
Please open a issue
